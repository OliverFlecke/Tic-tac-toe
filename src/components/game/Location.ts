/**
 * Represents a location in the game
 */
export default class Location {

    private _x : number;
    public get x() : number {
        return this._x;
    }
    public set x(v : number) {
        this._x = v;
    }

    private _y : number;
    public get y() : number {
        return this._y;
    }
    public set y(v : number) {
        this._y = v;
    }

    constructor(x: number = 0, y: number = 0) {
        this._x = x;
        this._y = y;
    }

    public toString() {
        return `(${this.x}, ${this.y})`;
    }
}
