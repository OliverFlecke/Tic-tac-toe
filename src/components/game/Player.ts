enum Player {
    Nought = 'O',
    Cross = 'X',
    None = ' '
}

export default Player;
