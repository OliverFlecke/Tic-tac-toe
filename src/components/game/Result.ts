enum Result {
    Noughts = 'Noughts',
    Crosses = 'Crosses',
    Tie = 'Tie',
    None = 'None'
}

export default Result;
