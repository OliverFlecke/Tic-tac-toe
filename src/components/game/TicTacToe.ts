import Location from './Location';
import Player from './Player';
import Result from './Result';

export default class TicTacToe {

    private _currentPlayer: Player = Player.Cross;

    private _grid: Player[][] = [
        Array(3).fill(Player.None),
        Array(3).fill(Player.None),
        Array(3).fill(Player.None)
    ];

    /**
     * Creates a deep clone of the TicTacToe object
     * @returns A deep clone of the object
     */
    public clone(): TicTacToe {
        const game = new TicTacToe();

        this.grid.forEach((row, y) => {
            row.forEach((cell, x) => game.setLocation(x, y, cell));
        });
        game._currentPlayer = this._currentPlayer;

        return game;
    }

    /**
     * @returns The game grid
     */
    public get grid() {
        return this._grid;
    }

    /**
     * Resets the game
     */
    public reset() {
        this._grid.forEach((row, y) => {
            row.forEach((_, x) => this.clearLocation(x, y));
        });
        this._currentPlayer = Player.Cross;
    }

    /**
     * Clear the given cell
     * @param x The x-coordinate of the cell to clear
     * @param y The y-coordinate of the cell to clear
     */
    private clearLocation(x: number, y: number) {
        this._grid[y][x] = Player.None;
    }

    /**
     * @param x The x-coordinate to set
     * @param y The y-coordinate to set
     * @param player The player for whom the position should be marked
     */
    private setLocation(x: number, y: number, player: Player) {
        if (this._grid[y][x] === Player.None) {
            this._grid[y][x] = player;
        }
    }

    /**
     * @param x The x-coordinate of the position to get
     * @param y The y-coordinate of the position to get
     * @returns The player token at the given position
     */
    public getLocation(x: number, y: number): Player {
        return this._grid[y][x];
    }

    /**
     * @param x The x-coordinate to do the action
     * @param y The y-coordinate to do the action
     * @param player The player doing the action
     */
    public doAction(x: number, y: number, player: Player = this.currentPlayer) {
        if (this.noMorePossibleMoves()) throw Error('There is no more possible moves. The game has ended');
        if (this._currentPlayer !== player) throw Error('Moves made by players have to alternate');
        if (this.getLocation(x, y) !== Player.None) throw Error('The location is already occupied');

        this._currentPlayer = this.nextPlayer;
        this.setLocation(x, y, player);
    }

    /**
     * @returns The current player of the game
     */
    public get currentPlayer(): Player {
        return this._currentPlayer;
    }

    /**
     * Find the next player to take a turn by alternating the players
     * @returns The next player to take a turn
     */
    public get nextPlayer(): Player {
        return this._currentPlayer === Player.Cross ? Player.Nought : Player.Cross;
    }

    /**
     * @returns The winner, if the player has three in a row horizontally
     */
    private checkRows(): Player | undefined {
        let value;
        for (const key in this.grid) {
            if (this.grid.hasOwnProperty(key)) {
                const line = this.grid[key];
                value = line.reduce((a, b) => a === b ? a : Player.None);
                if (value !== Player.None) {
                    return value;
                }
            }
        }

        return;
    }

    /**
     * @returns The winner, if the player has three in a row vertically
     */
    private checkColumns(): Player | undefined {
        let value;
        for (let x = 0; x < 3; x++) {
            value = this.grid.reduce((a, b) => a[x] === b[x] ? a : Array(3).fill(Player.None));
            if (value[x] !== Player.None) {
                return value[x];
            }
        }

        return;
    }

    /**
     * @returns The winner, if the player has three in a row in the diagonal
     */
    private checkDiagonal(): Player | undefined {
        if (this.getLocation(0, 0) === this.getLocation(1, 1)
            && this.getLocation(1, 1) === this.getLocation(2, 2)) {
            return this.getLocation(1, 1);
        }

        if (this.getLocation(2, 0) === this.getLocation(1, 1)
            && this.getLocation(1, 1) === this.getLocation(0, 2)) {
            return this.getLocation(1, 1);
        }

        return undefined;
    }

    /**
     * @returns The player whom is the winner of the game
     */
    public get winner(): Player {
        return this.checkColumns() ||
            this.checkRows() ||
            this.checkDiagonal() ||
            Player.None;
    }

    /**
     * @returns True if the game has ended
     */
    public hasEnded(): boolean {
        return this.result !== Result.None;
    }

    /**
     * @returns True, if there are no more possible moves
     */
    public noMorePossibleMoves(): boolean {
        return this.grid.every(row => row.every(cell => cell !== Player.None));
    }

    /**
     * @returns The result of the game
     */
    public get result(): Result {
        const winner = this.winner;
        if (winner === Player.Cross) return Result.Crosses;
        if (winner === Player.Nought) return Result.Noughts;

        if (this.noMorePossibleMoves()) return Result.Tie;

        return Result.None;
    }

    /**
     * @returns A string representation of the game
     */
    public toString(): string {
        return this.grid.map(line => line.join('|')).join('\n');
    }

    /**
     * @param game The game to find the free locations in
     * @returns A list of free locations in the given game
     */
    public freeLocations(): Location[] {
        const locations: Location[] = [];

        this.grid.forEach((row, y) => {
            row.forEach((cell, x) => {
                if (cell === Player.None) locations.push(new Location(x, y));
            });
        });

        return locations;
    }
}
