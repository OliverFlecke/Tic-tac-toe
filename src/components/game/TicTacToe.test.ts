import Location from './Location';
import Player from './Player';
import Result from './Result';
import TicTacToe from './TicTacToe';

const completeGame = () => {
    const game = new TicTacToe();
    game.doAction(2, 2, Player.Cross);
    game.doAction(0, 0, Player.Nought);
    game.doAction(2, 0, Player.Cross);
    game.doAction(1, 0, Player.Nought);
    game.doAction(0, 1, Player.Cross);
    game.doAction(1, 1, Player.Nought);
    game.doAction(2, 1, Player.Cross);
    game.doAction(0, 2, Player.Nought);
    game.doAction(1, 2, Player.Cross);

    return game;
};

describe('Basic tests of the game.', () => {
    test('Can render the empty game', () => {
        const game = new TicTacToe();
        expect(game.toString()).toEqual(' | | \n | | \n | | ');
    });

    test('Can render a game with Xs and Os', () => {
        const game = new TicTacToe();
        game.doAction(1, 1, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        expect(game.toString()).toEqual('O| | \n |X| \n | | ');
    });

    test('Can render a game with Xs and Os', () => {
        const game = new TicTacToe();
        game.doAction(2, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        game.doAction(2, 0, Player.Cross);
        game.doAction(1, 0, Player.Nought);
        game.doAction(0, 1, Player.Cross);
        game.doAction(1, 1, Player.Nought);
        game.doAction(2, 1, Player.Cross);
        game.doAction(0, 2, Player.Nought);
        game.doAction(1, 2, Player.Cross);
        expect(game.toString()).toEqual('O|O|X\nX|O|X\nO|X|X');
    });
});

describe('Game can check if there are any more possible moves', () => {
    test('A new game', () => {
        const game = new TicTacToe();
        expect(game.noMorePossibleMoves()).toBeFalsy();
    });
    test('A game with some amount of moves', () => {
        const game = new TicTacToe();
        game.doAction(0, 1, Player.Cross);
        game.doAction(1, 1, Player.Nought);
        game.doAction(2, 1, Player.Cross);
        game.doAction(0, 2, Player.Nought);
        game.doAction(1, 2, Player.Cross);
        expect(game.noMorePossibleMoves()).toBeFalsy();
    });
    test('A game with all cells occupied', () => {
        const game = completeGame();
        expect(game.noMorePossibleMoves()).toBeTruthy();
    });
});

describe('Can find the result of the game', () => {

    test('Can find the winner of the game horizontally', () => {
        for (let y = 0; y < 3; y++) {
            // Setup
            const game = new TicTacToe();
            game.doAction(0, y, Player.Cross);
            game.doAction(0, y ? 0 : y + 1, Player.Nought);
            game.doAction(1, y, Player.Cross);
            game.doAction(1, y ? 0 : y + 1, Player.Nought);
            game.doAction(2, y, Player.Cross);

            // Assert
            expect(game.winner).toEqual(Player.Cross);
        }
    });

    test('Can find the winner of the game vertically', () => {
        for (let x = 0; x < 3; x++) {
            // Setup
            const game = new TicTacToe();
            game.doAction(x, 0, Player.Cross);
            game.doAction(x ? 0 : x + 1, 0, Player.Nought);
            game.doAction(x, 1, Player.Cross);
            game.doAction(x ? 0 : x + 1, 1, Player.Nought);
            game.doAction(x, 2, Player.Cross);

            // Assert
            expect(game.winner).toEqual(Player.Cross);
        }
    });

    test('Can find the winner in the diagonal', () => {
        // Setup
        const game = new TicTacToe();
        for (let i = 0; i < 3; i++) {
            game.doAction(i, i, Player.Cross);
            if (i === 0) game.doAction(2, 0, Player.Nought);
            if (i === 1) game.doAction(2, 1, Player.Nought);
        }

        // Assert
        expect(game.toString()).toEqual('X| |O\n |X|O\n | |X');
        expect(game.winner).toEqual(Player.Cross);
    });

    test('Can find the winner in the second diagonal', () => {
        // Setup
        const game = new TicTacToe();
        for (let i = 0; i < 3; i++) {
            game.doAction(2 - i, i, Player.Cross);
            if (i === 0) game.doAction(0, 0, Player.Nought);
            if (i === 1) game.doAction(0, 1, Player.Nought);
        }

        // Assert
        expect(game.toString()).toEqual('O| |X\nO|X| \nX| | ');
        expect(game.winner).toEqual(Player.Cross);
    });

    test('The game is a tie if all cells are occupied', () => {
        const game = new TicTacToe();

        game.doAction(2, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        game.doAction(2, 0, Player.Cross);
        game.doAction(2, 1, Player.Nought);
        game.doAction(0, 1, Player.Cross);
        game.doAction(1, 1, Player.Nought);
        game.doAction(1, 0, Player.Cross);
        game.doAction(0, 2, Player.Nought);
        game.doAction(1, 2, Player.Cross);

        expect(game.toString()).toMatchSnapshot();
        expect(game.noMorePossibleMoves()).toBeTruthy();
        expect(game.result).toEqual(Result.Tie);
    });

    test('Game has no more possible moves, but there is a winner', () => {
        const game = new TicTacToe();

        game.doAction(2, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        game.doAction(2, 0, Player.Cross);
        game.doAction(1, 0, Player.Nought);
        game.doAction(0, 1, Player.Cross);
        game.doAction(1, 1, Player.Nought);
        game.doAction(2, 1, Player.Cross);
        game.doAction(0, 2, Player.Nought);
        game.doAction(1, 2, Player.Cross);

        expect(game.toString()).toMatchSnapshot();
        expect(game.noMorePossibleMoves()).toBeTruthy();
        expect(game.result).toEqual(Result.Crosses);
    });
});

describe('Doing actions in a round of the game.', () => {
    test('Crosses have to start', () => {
        const game = new TicTacToe();

        game.doAction(0, 0, Player.Cross);

        expect(game.grid[0][0]).toEqual(Player.Cross);
    });

    test('Noughts cannot make the first move', () => {
        const game = new TicTacToe();

        function doAction() {
            game.doAction(0, 0, Player.Nought);
        }
        expect(doAction).toThrowErrorMatchingSnapshot();
    });

    test('Noughts cannot make two moves in a row', () => {
        const game = new TicTacToe();

        game.doAction(0, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        function doAction() {
            game.doAction(0, 1, Player.Nought);
        }

        expect(game.getLocation(0, 0)).toEqual(Player.Nought);
        expect(doAction).toThrowErrorMatchingSnapshot();
    });

    test('Crosses cannot make two moves in a row', () => {
        const game = new TicTacToe();

        game.doAction(0, 0, Player.Cross);
        function doAction() {
            game.doAction(0, 1, Player.Cross);
        }

        expect(game.getLocation(0, 0)).toEqual(Player.Cross);
        expect(doAction).toThrowErrorMatchingSnapshot();
    });

    test('An action cannot be taken on a spot that is occupied', () => {
        // Setup
        const game = new TicTacToe();
        game.doAction(0, 0, Player.Cross);

        // Test function
        function doAction() {
            game.doAction(0, 0, Player.Nought);
        }

        expect(game.getLocation(0, 0)).toEqual(Player.Cross);
        expect(doAction).toThrowErrorMatchingSnapshot();
        expect(game.getLocation(0, 0)).toEqual(Player.Cross);
    });

    test('It is not possible to make a move when all cells are field', () => {
        const game = new TicTacToe();

        function doAction() {
            game.doAction(0, 0, Player.Nought);
        }

        game.doAction(2, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        game.doAction(2, 0, Player.Cross);
        game.doAction(1, 0, Player.Nought);
        game.doAction(0, 1, Player.Cross);
        game.doAction(1, 1, Player.Nought);
        game.doAction(2, 1, Player.Cross);
        game.doAction(0, 2, Player.Nought);
        game.doAction(1, 2, Player.Cross);

        expect(game.toString()).toMatchSnapshot();
        expect(game.noMorePossibleMoves()).toBeTruthy();
        expect(doAction).toThrowErrorMatchingSnapshot();
    });
});

describe('Testing agent', () => {
    test('finding no free cells in a finished game', () => {
        const game = new TicTacToe();

        game.doAction(2, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        game.doAction(2, 0, Player.Cross);
        game.doAction(1, 0, Player.Nought);
        game.doAction(0, 1, Player.Cross);
        game.doAction(1, 1, Player.Nought);
        game.doAction(2, 1, Player.Cross);
        game.doAction(0, 2, Player.Nought);
        game.doAction(1, 2, Player.Cross);

        const freeLocations = game.freeLocations();

        expect(freeLocations.length).toEqual(0);
    });

    test('finding all the free cells in a game', () => {
        const game = new TicTacToe();

        game.doAction(2, 2, Player.Cross);
        game.doAction(0, 0, Player.Nought);
        game.doAction(2, 0, Player.Cross);
        game.doAction(1, 0, Player.Nought);
        game.doAction(0, 1, Player.Cross);

        const freeLocations = game.freeLocations();

        expect(freeLocations).toHaveLength(4);
        expect(freeLocations).toContainEqual(new Location(1, 2));
        expect(freeLocations).toContainEqual(new Location(1, 1));
        expect(freeLocations).toContainEqual(new Location(2, 1));
        expect(freeLocations).toContainEqual(new Location(0, 2));
    });

    test('find all cells in a new game', () => {
        const game = new TicTacToe();
        const freeLocations = game.freeLocations();

        // Check for free locations
        expect(freeLocations).toHaveLength(9);
        expect(freeLocations).toContainEqual(new Location(0, 0));
        expect(freeLocations).toContainEqual(new Location(0, 1));
        expect(freeLocations).toContainEqual(new Location(0, 2));
        expect(freeLocations).toContainEqual(new Location(1, 0));
        expect(freeLocations).toContainEqual(new Location(1, 1));
        expect(freeLocations).toContainEqual(new Location(1, 2));
        expect(freeLocations).toContainEqual(new Location(2, 0));
        expect(freeLocations).toContainEqual(new Location(2, 1));
        expect(freeLocations).toContainEqual(new Location(2, 2));
    });
});

describe('Cloning a game', () => {
    test('creating a simple deep clone', () => {
        const game = new TicTacToe();

        const clone = game.clone();

        expect(clone).not.toBe(game);
        expect(clone.grid).not.toBe(game.grid);
        expect(clone.grid).toEqual(game.grid);
        expect(clone.currentPlayer).toEqual(game.currentPlayer);

        clone.doAction(0, 0, Player.Cross);

        expect(clone.getLocation(0, 0)).toEqual(Player.Cross);
        expect(game.getLocation(0, 0)).toEqual(Player.None);
    });
});
