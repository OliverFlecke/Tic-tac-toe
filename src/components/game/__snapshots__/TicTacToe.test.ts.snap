// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can find the result of the game Game has no more possible moves, but there is a winner 1`] = `
"O|O|X
X|O|X
O|X|X"
`;

exports[`Can find the result of the game The game is a tie if all cells are occupied 1`] = `
"O|X|X
X|O|O
O|X|X"
`;

exports[`Doing actions in a round of the game. An action cannot be taken on a spot that is occupied 1`] = `"The location is already occupied"`;

exports[`Doing actions in a round of the game. Crosses cannot make two moves in a row 1`] = `"Moves made by players have to alternate"`;

exports[`Doing actions in a round of the game. It is not possible to make a move when all cells are field 1`] = `
"O|O|X
X|O|X
O|X|X"
`;

exports[`Doing actions in a round of the game. It is not possible to make a move when all cells are field 2`] = `"There is no more possible moves. The game has ended"`;

exports[`Doing actions in a round of the game. Noughts cannot make the first move 1`] = `"Moves made by players have to alternate"`;

exports[`Doing actions in a round of the game. Noughts cannot make two moves in a row 1`] = `"Moves made by players have to alternate"`;
